core/src/main/java/org/apache/iceberg/actions/SizeBasedDataRewriter.java

```diff
@@ -84,13 +87,34 @@ private boolean shouldRewrite(List<FileScanTask> group) {
     return enoughInputFiles(group)
         || enoughContent(group)
         || tooMuchContent(group)
-        || anyTaskHasTooManyDeletes(group);
+        || anyTaskHasTooManyDeletes(group)
+        || anyTaskHasTooHighDeleteRatio(group);
   }
 
   private boolean anyTaskHasTooManyDeletes(List<FileScanTask> group) {
     return group.stream().anyMatch(this::tooManyDeletes);
   }
 
+  private boolean anyTaskHasTooHighDeleteRatio(List<FileScanTask> group) {
+    return group.stream().anyMatch(this::tooHighDeleteRatio);
+  }
+
+  private boolean tooHighDeleteRatio(FileScanTask task) {
+    if (null == task.deletes() || task.deletes().isEmpty()) {
+      return false;
+    }
+
+    if (ContentFileUtil.containsSingleDV(task.deletes())
+        || task.deletes().stream().allMatch(ContentFileUtil::isFileScoped)) {
+      long sum = task.deletes().stream().mapToLong(ContentFile::recordCount).sum();
+      double deletedRecords = (double) Math.min(sum, task.file().recordCount());
```

Comment:
Does this line address an edge case?

Reply:
yes, this addressed an edge case that is being tested by `testDataFilesRewrittenWithMaxDeleteRatio` where we produce more positions to be deleted than actual records